### PR TITLE
fix: Hasura permissions

### PR DESCRIPTION
**Pull request by igorosip0v**
Noticed fixes while adjusting code for Hasura changes

### CODE REVIEW

File: hasura/metadata/databases/default/tables/public_membership.yaml [hunk 1]
```diff
@@ -56,3 +57,11 @@ delete_permissions:
...
     permission:
       filter: {}
+  - role: user
+    permission:
+      filter:
+        _and:
+          - team_id:
+              _eq: X-Hasura-Team-Id
+          - role:
+              _eq: OWNER
```
Comment (andy-t-wang) [inline hasura/metadata/databases/default/tables/public_membership.yaml:67]: This means rows where team id matches and has a member that is an owner? Need the user-id here in an and with the role?

**Review by andy-t-wang — CHANGES_REQUESTED** [review]
Also looks like some permissions still have the permission nesting issues from #453 where it's not checking the user's role specifically but just that the role exists

File: hasura/metadata/databases/default/tables/public_membership.yaml (hunk 1 above)
Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_membership.yaml:70]: Any OWNER in this team, as I can see it.

File: hasura/metadata/databases/default/tables/public_user.yaml [hunk 2]
```diff
@@ -67,11 +70,3 @@ update_permissions:
         id:
           _eq: X-Hasura-User-Id
       check: null
-delete_permissions:
```
Comment (pdtfh) [inline hasura/metadata/databases/default/tables/public_user.yaml:70]: ok for now but let's please note internally we need to add/confirm full support for: deleting your own user account; orphaned users who don't have a team (frontend handling mainly).

File: hasura/metadata/databases/default/tables/public_user.yaml [hunk 3]
```diff
@@ -48,8 +50,9 @@ select_permissions:
         - is_subscribed
         - name
       filter:
-        id:
-          _eq: X-Hasura-User-Id
+        memberships:
```
Comment (pdtfh) [inline hasura/metadata/databases/default/tables/public_user.yaml:53]: is there a way we don't expose auth0Id and is_subscribed other than to yourself?

File: hasura/metadata/databases/default/tables/public_user.yaml [hunk 4]
```diff
@@ -23,6 +23,7 @@ insert_permissions:
         - ironclad_id
         - is_subscribed
         - name
+        - team_id
```
Comment (pdtfh) [inline hasura/metadata/databases/default/tables/public_user.yaml:26]: is this for temporary compatibility?

File: hasura/metadata/databases/default/tables/public_membership.yaml (hunk 1 above)
Comment (andy-t-wang) [reply hasura/metadata/databases/default/tables/public_membership.yaml:70]: But does this check if the user submitting the delete request is the Owner? 
Think it's missing a 
```
- user_id:
     _eq: X-Hasura-User-Id
```

Comment (andy-t-wang) [reply hasura/metadata/databases/default/tables/public_membership.yaml:70]: For some reason I keep reading this as it filters for rows where you are on the team and the role is owner. So instead of it saying you are on the team and the owner so you can delete any role on the team.
It's saying you are on the team and the role you are deleting is an owner role

**Comment by andy-t-wang** [comment]
Looks good to me besides that delete permission public_membership.yaml.

File: hasura/metadata/databases/default/tables/public_user.yaml (hunk 2 above)
Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_user.yaml:70]: @pdtfh First we need to discuss how it will look like on the UI side. Because now we will have two different processes:
1. Removing user from the team (Deleting `membership` table row)
2. Deleting user from the Dev Portal completely (Deleting `user` table row)

File: hasura/metadata/databases/default/tables/public_api_key.yaml [hunk 5]
```diff
@@ -47,13 +45,13 @@ select_permissions:
...
-              _eq: X-Hasura-Team-Id
-          - team:
-              memberships:
-                user_id:
+        team:
+          memberships:
+            _and:
+              - user_id:
                   _eq: X-Hasura-User-Id
+              - team_id:
```
Comment (pdtfh) [inline hasura/metadata/databases/default/tables/public_api_key.yaml:53]: let's only let ADMIN & OWNER read API keys

File: hasura/metadata/databases/default/tables/public_membership.yaml [hunk 6]
```diff
@@ -56,3 +60,11 @@ delete_permissions:
   - role: service
     permission:
       filter: {}
+  - role: user
```
Comment (pdtfh) [inline hasura/metadata/databases/default/tables/public_membership.yaml:63]: can be in a separate PR, but we should let ADMINs change all the way to ADMIN (including downgrading)

Comment (pdtfh) [reply hasura/metadata/databases/default/tables/public_membership.yaml:63]: same for deletion below

**Review by pdtfh — APPROVED** [review]

File: hasura/metadata/databases/default/tables/public_user.yaml (hunk 2 above)
Comment (pdtfh) [reply hasura/metadata/databases/default/tables/public_user.yaml:70]: 👍 
1. Should be from the list of teammates.
2. Should only be done by each person, can be done from profile settings. We also need to handle the case of you logging in and not belonging to any team. You should be able to create a new team or delete your account. (CC @andy-t-wang to make sure it's covered in design).

File: hasura/metadata/databases/default/tables/public_membership.yaml (hunk 1 above)
Comment (pdtfh) [reply hasura/metadata/databases/default/tables/public_membership.yaml:70]: I think you're right @andy-t-wang, this seems to be looking at the role in the row, not the role of the user

Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_membership.yaml:70]: > It's saying you are on the team and the role you are deleting is an owner role
Just checked, seems you are right. But the problem is that if we will add check for `X-Hasura-User-Id` it will allow remove only the rows with you as member.

In this case I think we need to leave only 
```
- team_id
  _eq: X-Hasura-Team-Id
 ```

And then just show Remove button only for the ADMIN and OWNER roles from the code. This is already implemented in #457.

Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_membership.yaml]: @pdtfh @andy-t-wang Please check [this](https://github.com/worldcoin/developer-portal/pull/456/commits/a61ac17852622c5b5cd639cd8a69483e1de49589) commit

File: hasura/metadata/databases/default/tables/public_user.yaml (hunk 3 above)
Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_user.yaml:53]: Not sure about this with the single `user` hasura role.

File: hasura/metadata/databases/default/tables/public_team.yaml [hunk 7]
```diff
@@ -56,41 +56,36 @@ select_permissions:
...
-              team_id:
-                _eq: X-Hasura-Team-Id
-          - memberships:
-              user_id:
+        memberships:
+          _and:
+            - user_id:
                 _eq: X-Hasura-User-Id
+            - team_id:
+                _eq: X-Hasura-Team-Id
```
Comment (igorosip0v) [inline hasura/metadata/databases/default/tables/public_team.yaml:64]: @andy-t-wang @pdtfh Just found out that we need to remove this team_id check from select row check in team table for user role. Otherwise, we can't pick a list of teams for teams selector.

I think it's ok, user should be able select all the teams related with his x-hasura-user-id. Or am I missing something? 

I will push this change when I finish team selector

File: hasura/metadata/databases/default/tables/public_user.yaml (hunk 3 above)
Comment (pdtfh) [reply hasura/metadata/databases/default/tables/public_user.yaml:53]: ok let's track this in a Linear issue to revisit

File: hasura/metadata/databases/default/tables/public_team.yaml (hunk 7 above)
Comment (pdtfh) [reply hasura/metadata/databases/default/tables/public_team.yaml:64]: yes, aligned on this

File: hasura/metadata/databases/default/tables/public_membership.yaml (hunk 1 above)
Comment (pdtfh) [reply hasura/metadata/databases/default/tables/public_membership.yaml]: I don't think this is the best approach, permissions as they currently are would allow any teammate to delete a membership record. for update permissions, I believe we're still letting any team member update any record that belongs to an OWNER, regardless of the permission level of the user executing.

In addition, do we plan to validate the `x-hasura-team-id` in the `api/v1/graphql` endpoint ? because in other models we rely on permissions to validate this, but here you could have an incorrect value and bypass permissions

File: hasura/metadata/databases/default/tables/public_membership.yaml [hunk 8]
```diff
@@ -30,10 +30,14 @@ select_permissions:
       filter: {}
   - role: user
     permission:
-      columns: []
+      columns:
+        - role
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_membership.yaml:35]: Yes. Yes

File: hasura/metadata/databases/default/tables/public_user.yaml (hunk 3 above)
Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_user.yaml:53]: [WID-657](https://linear.app/worldcoin/issue/WID-657/make-auth0id-and-is-subscribed-fields-visible-only-for-yourself)

File: hasura/metadata/databases/default/tables/public_api_key.yaml (hunk 5 above)
Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_api_key.yaml:53]: Done

File: hasura/metadata/databases/default/tables/public_membership.yaml (hunk 1 above)
Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_membership.yaml]: > I don't think this is the best approach, permissions as they currently are would allow any teammate to delete a membership record. 
How MEMBER can do this if he has nothing to click to invoke deletion? 

As additional measures I can add checks for roles: `{_or: [{role: {_eq: MEMBER}, {role: {_eq: ADMIN}}]}`. This way there is no possibility to remove OWNER.

> In addition, do we plan to validate the x-hasura-team-id in the api/v1/graphql endpoint ?
Can you clarify?

File: hasura/metadata/databases/default/tables/public_user.yaml (hunk 2 above)
Comment (igorosip0v) [reply hasura/metadata/databases/default/tables/public_user.yaml]: > Should only be done by each person

Updated. Add check `{id: {_eq: X-Hasura-User-Id}}`
